// kram/dist/default/plugin/minify.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _htmlMinifier = require('html-minifier');

var _htmlMinifier2 = _interopRequireDefault(_htmlMinifier);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var minifyEngine = _htmlMinifier2.default.minify; /**
                                                   * @file html 压缩插件
                                                   * @author tanglei (tanglei02@baidu.com)
                                                   */

var Minify = function () {
    function Minify() {
        var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
            priority = _ref.priority;

        (0, _classCallCheck3.default)(this, Minify);

        this.priority = priority;
    }

    (0, _createClass3.default)(Minify, [{
        key: 'apply',
        value: function apply(on, app) {
            on(app.module.hook.STAGES.AFTER_PARSE, function (html) {
                try {
                    return minifyEngine(html, {
                        collapseWhitespace: true,
                        minifyCSS: true
                    });
                } catch (e) {
                    return html;
                }
            }, this.priority);
        }
    }]);
    return Minify;
}();

exports.default = Minify;
