// kram/dist/module/hook/stage.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.STAGE_SET = exports.STAGES = exports.RENDER_STAGES = exports.DONE = exports.RENDER_PREFIX = exports.CREATE_MENU = exports.CREATE_DOC_STORE_OBJECT = exports.AFTER_PARSE = exports.BEFORE_PARSE = exports.AFTER_FILE_PROCESS = exports.GET_CHANGED_ENTRY_FILES = exports.GET_CHANGED_FILES = exports.GET_ORIGINAL_FILES = exports.AFTER_LOAD = exports.BEFORE_LOAD = exports.START = undefined;

var _renderer = require('../parser/renderer');

var START = exports.START = 'start'; /**
                                      * @file 事件钩子定义
                                      * @author tanglei (tanglei02@baidu.com)
                                      */

var BEFORE_LOAD = exports.BEFORE_LOAD = 'beforeLoad';
var AFTER_LOAD = exports.AFTER_LOAD = 'afterLoad';

var GET_ORIGINAL_FILES = exports.GET_ORIGINAL_FILES = 'getOriginalFiles';
var GET_CHANGED_FILES = exports.GET_CHANGED_FILES = 'getChangedFiles';
var GET_CHANGED_ENTRY_FILES = exports.GET_CHANGED_ENTRY_FILES = 'getChangedEntryFiles';
var AFTER_FILE_PROCESS = exports.AFTER_FILE_PROCESS = 'afterFileProcess';

var BEFORE_PARSE = exports.BEFORE_PARSE = 'beforeParse';
var AFTER_PARSE = exports.AFTER_PARSE = 'afterParse';
var CREATE_DOC_STORE_OBJECT = exports.CREATE_DOC_STORE_OBJECT = 'createDocStoreObject';

var CREATE_MENU = exports.CREATE_MENU = 'createMenu';

var RENDER_PREFIX = exports.RENDER_PREFIX = 'render:';

var DONE = exports.DONE = 'done';

var RENDER_STAGES = exports.RENDER_STAGES = _renderer.RENDER_NAMES.reduce(function (res, name) {
    res['RENDER_' + name.toUpperCase()] = RENDER_PREFIX + name;
    return res;
}, {});

var STAGES = exports.STAGES = Object.assign({
    START: START,

    BEFORE_LOAD: BEFORE_LOAD,
    AFTER_LOAD: AFTER_LOAD,

    GET_ORIGINAL_FILES: GET_ORIGINAL_FILES,
    GET_CHANGED_FILES: GET_CHANGED_FILES,
    GET_CHANGED_ENTRY_FILES: GET_CHANGED_ENTRY_FILES,
    AFTER_FILE_PROCESS: AFTER_FILE_PROCESS,

    BEFORE_PARSE: BEFORE_PARSE,
    AFTER_PARSE: AFTER_PARSE,
    CREATE_DOC_STORE_OBJECT: CREATE_DOC_STORE_OBJECT,

    CREATE_MENU: CREATE_MENU,

    DONE: DONE
}, RENDER_STAGES);

var STAGE_SET = exports.STAGE_SET = Object.keys(STAGES).reduce(function (set, key) {
    return set.add(STAGES[key]);
}, new Set());
