// kram/dist/default/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

exports.default = function (app) {
    var defaultConfig = Object.keys(config).reduce(function (res, key) {
        if (typeof config[key] === 'function') {
            res[key] = config[key](app);
        } else {
            res[key] = config[key];
        }

        return res;
    }, {});

    return {
        config: defaultConfig,
        module: _module
    };
};

var _store = require('./config/store');

var _store2 = _interopRequireDefault(_store);

var _highlight = require('./config/highlight');

var _highlight2 = _interopRequireDefault(_highlight);

var _parser = require('./config/parser');

var _parser2 = _interopRequireDefault(_parser);

var _plugin = require('./config/plugin');

var _plugin2 = _interopRequireDefault(_plugin);

var _loader = require('./config/loader');

var _loader2 = _interopRequireDefault(_loader);

var _downloadGitRepo = require('./loader/download-git-repo');

var _local = require('./loader/local');

var _npm = require('./highlight/npm');

var _vue = require('./highlight/vue');

var _insert = require('./plugin/insert');

var _insert2 = _interopRequireDefault(_insert);

var _minify = require('./plugin/minify');

var _minify2 = _interopRequireDefault(_minify);

var _style = require('./plugin/style');

var _style2 = _interopRequireDefault(_style);

var _stylus = require('./plugin/stylus');

var _stylus2 = _interopRequireDefault(_stylus);

var _meta = require('./plugin/meta');

var _meta2 = _interopRequireDefault(_meta);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file 默认配置文件
 * @author tangei (tanglei02@baidu.com)
 */

var logger = console;

var config = {
    store: _store2.default,
    logger: logger,
    highlight: _highlight2.default,
    parser: _parser2.default,
    plugin: _plugin2.default,
    loader: _loader2.default
};

var _module = {
    loader: {
        downloadGitRepo: _downloadGitRepo.downloadGitRepo,
        local: _local.local
    },
    highlight: {
        npm: _npm.npm,
        vue: _vue.vue
    },
    store: _store2.default,
    plugin: {
        Insert: _insert2.default,
        Minify: _minify2.default,
        Style: _style2.default,
        Stylus: _stylus2.default,
        Meta: _meta2.default
    }
};
